docs/proposals/20250513-propogate-taints.md

```diff
@@ -0,0 +1,455 @@
+
+### Risks and Mitigations
+
+Managing the `Taint`s on `Node`s is considered a highly privileged action in Kubernetes; it even has its own top level `kubectl taint` command.
+The Kubernetes scheduler uses taints rather than `Conditions` to decide when to evict workloads.
+Updating these in place could then evict workloads unintentionally, or disrupt other systems that rely on taints being present.
+
+This risk can be mitigated by ensuring Cluster API only modifies taints that it owns on nodes, as decribed in the above.
+
+## Alternatives
+
+Deciding whether or not to reconcile taint changes continuously has been a challenge for the Cluster API.
+Historically, the v1alpha1 API included a `Machine.Taints` field.
+However, since this field was mostly used in cluster bootstrapping, it was later extracted into bootstrap provider implementations.
+There is no standard API in bootstrap provider's though for taints.
+
+Moving forward, two broad alternatives that have been explored in light of this: adding taints only at bootstrap time, and making taints immutable at the `Machine` level. Both methods would require that a Node be replaced in order to make any changes to the taints.
+
+While this simplifies the implementation logic for Cluster API, it may be surprising to many users, since the Kubernetes documentation presents taints as a mutable field on a Node.
+This would also mean that there are two different behaviors when modifying metadata within Cluster API, which could again be very confusing.
+There is already precedent for leaving infrastructure in-place when Kubernetes-only fields are modified, and this proposal seeks to align with the established function.
+
+## Upgrade Strategy
+
+Taint support is a net-new field, and therefore must be optional and not affect upgrades.
+
+## Additional Details
+
+[^1]: It is worth noting that there has been discussion about making the taints on a Node a "map" list type and allowing for ownership of individual taints.
+As of this writing, the [pull request](https://github.com/kubernetes/kubernetes/pull/128866) and [issue](https://github.com/kubernetes/kubernetes/issues/117142) remain open.
+This proposal should be unaffected regardless of any upstream change to the handling of taints, except that using a "map" type would simplify the implementation and allow us to cooperate with other field managers.
+
+
+### Test Plan [optional]
+
+- There should be proper unit and integration test coverage when introducing the changes.
+- An e2e test should validate the high-level user-experience
+  - Usage of an `Initialize` and `Always` taint
+  - Adding or removing an `Always` and `Initialize` taint on day 2
+
+### Graduation Criteria [optional]
+
+Open question: Should we feature-gate this new field?
```

Comment:
Yes, please.

> Considering that "not setting the fields" should be equivalent to "disabling the feature gate" seems to be good enough.

The feature gate would give us some time to get experience with this feature and make small adjustments if necessary.

I would expect us to use the feature gate to block setting the fields via webhooks and to gate the corresponding logic in controllers

Reply:
Adjusted the section to mention the above points.

Reply:
Also the paragraph at the end proposes a feature gate name.